File: AnalysisCode/systematics/rebinoutput.cc
// inlcude c++ library classes
#include <string>
#include <vector>
#include <exception>
#include <iostream>

// include ROOT classes
#include "TROOT.h" 
#include "TH1D.h"
#include "TFile.h"
#include "TTree.h"

// include other parts of framework
#include "../../Tools/interface/stringTools.h"
#include "../../Tools/interface/QuantileBinner.h"

#include "systematicsTools.h"

int main( int argc, char* argv[] ){

    std::vector< std::string > argvStr( &argv[0], &argv[0] + argc );

    if( argc != 8 ){
	std::cerr << "### ERROR ###: rebinoutput.cc requires 7 arguments to run: ";
	std::cerr << "background_file_path, input_file_path, output_file_path, variable_name,";
	std::cerr << " mode, lastBinCount, binRatio" << std::endl;;
	return -1;
    }
    
    // command line args:
    // - input file path: file to run on
    // - output file path: path to where the (rebinned) output should be placed
    // - variable_name: name of the variable to rebin
    // - mode: either "signal" or "background" (on which the rebinning will be based)
    // - eventsInLastBin and binRatio: rebinning parameters
    // - background_file_path: path containing all the backgrounds (needed for rebinning)
    //			       it can be the same as input_file_path if that file is 
    //			       already merged and contains all the backgrounds

    // parse arguments
    std::string bkg_file_path = argvStr[1];
    std::string input_file_path = argvStr[2];
    std::string output_file_path = argvStr[3];
    std::string variable_name = argvStr[4];
    std::string mode = argvStr[5];
    std::string lastBinCount = argvStr[6];
    double eventsInLastBin;
    bool useFraction = false;
    if(lastBinCount[lastBinCount.length()-1]!='f'){
	eventsInLastBin = std::stod(lastBinCount);
    }
    else{
	useFraction = true;
	eventsInLastBin = std::stod(lastBinCount.substr(0,lastBinCount.length()-1));
    }
    double binRatio = std::stod(argvStr[7]);

    // the input and background files are supposed to contain histograms with names of the form
    // <processTag>_<variableName>_<systematic>
    std::vector<std::string> mayNotContain;
    std::vector<std::string> mustContain;
    if(mode=="background"){
	mayNotContain = {"tZq","data"};
	mustContain = {variable_name, "nominal"};
    }
    else if(mode=="signal"){
	mayNotContain = {};
	mustContain = {variable_name,"nominal","tZq"};	
    }
    else{
	std::cerr << "### ERROR ###: mode '" << mode << "' not recognized." << std::endl;
	return -1;
    }

    // get all histograms from the background file
    std::vector<std::shared_ptr<TH1>> bkgHistogramVector = getHistogramsFromFile(bkg_file_path);
    // select the variable histograms from background
    bkgHistogramVector = selectHistograms(
        bkgHistogramVector, mustContain, mayNotContain );
    // add selected histograms
    std::shared_ptr<TH1> bkgHist = addHistograms( bkgHistogramVector );
    // recalculate events in last bin if fraction was speciied
    if( useFraction ){
	eventsInLastBin = eventsInLastBin*bkgHist->Integral();
    }
    
    // determine new bins using QuantileBinner
    std::vector<double> contents = exponentialBinningQuantiles( bkgHist.get(), 
	eventsInLastBin=eventsInLastBin, binRatio=binRatio );
    QuantileBinner rebinner( bkgHist.get(), contents );
    
    // print for testing
    //std::cout << "new content fractions: " << std::endl;
    //for(double content : contents){
    //	std::cout << " " << content;
    //}
    //std::cout << std::endl;

    // get all histograms from the input file
    std::vector<std::shared_ptr<TH1>> inHistogramVector = getHistogramsFromFile(input_file_path);
    // exclude previously rebinned histograms
    inHistogramVector = selectHistograms( inHistogramVector, std::vector<std::string>({}), 
                                      std::vector<std::string>({"rebinned"}) );
    // make rebinned histograms belonging to this variable
    std::vector<std::shared_ptr<TH1>> rebinnedHistogramVector;
    for( std::shared_ptr<TH1> histPtr: inHistogramVector ){
	if(!stringTools::stringContains(histPtr->GetName(),variable_name)) continue;
	std::shared_ptr<TH1> rebinnedHistogram = rebinner.rebinnedHistogram( histPtr.get() );
	// set new histogram name
	// note: make sure newvarname does not contain variable_name else replace will loop forever!
	std::string tempvarname = stringTools::removeOccurencesOf( variable_name, "fineBinned" );
	std::string newvarname = "_rebinned"+tempvarname.substr(1,std::string::npos);
	std::string name = stringTools::replace( std::string(histPtr->GetName()), 
						    variable_name, newvarname);
	rebinnedHistogram->SetName( name.c_str() );
	rebinnedHistogram->SetTitle( histPtr->GetTitle() );
	rebinnedHistogramVector.push_back( rebinnedHistogram );
    } 

    // open output file and write histograms to it
    TFile* f = TFile::Open(output_file_path.c_str(),"recreate");
    for(std::shared_ptr<TH1> histPtr : inHistogramVector){
	histPtr->Write();
    }
    for(std::shared_ptr<TH1> histPtr : rebinnedHistogramVector){
	histPtr->Write();
    }
    f->Close();
    return 0;
}

File: AnalysisCode/systematics/rebinoutput.py
#############################################################################
# script to automatically call the rebinner on the output of runsystematics #
#############################################################################
# after rebinning, envelopes are recalculated where needed and histograms are clipped

import os
import sys
import ROOT
sys.path.append(os.path.abspath('../../skimmer'))
from jobSubmission import submitQsubJob, initializeJobScript
from systematicstools import getfilestomerge
sys.path.append(os.path.abspath('../tools'))
import histtools as histtools

##### main function #####

def rebinoutput(topdir,year,region,npmode,mode,firstbincount,binfactor):

    # step 1: merge all relevant files using hadd
    # (the hadding here is only temporary, needed for rebinning)
    combpath = os.path.join(topdir,year+'combined',region,npmode)
    if os.path.exists(combpath):
        os.system('rm -r '+combpath)
    os.makedirs(combpath)
    filestomerge = getfilestomerge(topdir,year,region,npmode) # implicit printing
    if len(filestomerge)==0: return -1
    outputfile = os.path.join(combpath,'combined.root')
    print('running hadd command...')
    command = 'hadd '+outputfile
    for f in filestomerge: command += ' '+f
    os.system(command)

    ### for testing only ###
    #filestomerge = filestomerge[:3]

    # step 1b: for speed, remove all unnecessary histograms from the file
    # (only need nominal histograms for this specific variable)
    # note that the removing of histograms does not need to be very rigorous here,
    # the correct selection is done anyway in rebinoutput.cc
    histlist = histtools.loadallhistograms(outputfile)
    tempfilename = outputfile[:-5]+'_temp.root'
    if os.path.exists(tempfilename):
        os.system('rm '+tempfilename)
    tempfile = ROOT.TFile.Open(tempfilename,'recreate')
    for hist in histlist:
	if not ('_fineBinnedeventBDT' in hist.GetName() 
		and 'nominal' in hist.GetName()): continue
	hist.Write()
    tempfile.Close()
    os.system('mv {} {}'.format(tempfilename,outputfile))

    # step 2: rebin the discriminating variable
    print('rebinning...')
    if not os.path.exists('./rebinoutput'):
        print('### ERROR ###: rebinoutput executable does not seem to exist')
	return -1
    for f in filestomerge:
	print('now rebinning {}'.format(f))
	tempfilename = f[:-5]+'_temp.root'
	if os.path.exists(tempfilename):
	    os.system('rm '+tempfilename)
	command = './rebinoutput '+outputfile+' '+f+' '+tempfilename+' _fineBinnedeventBDT'
	command += ' '+mode+' '+firstbincount+' '+binfactor
	os.system(command)
	command = 'mv '+tempfilename+' '+f
	os.system(command)

    # step 3: recalculate envelopes and RMS's
    for f in filestomerge:
	print('now recalculating envelopes for {}'.format(f))
	gencommand = './reenvelope '+f+' '+f+' _rebinnedeventBDT'
	commands = []
	commands.append( gencommand + ' qcdScalesShapeVar qcdScalesShapeEnv envelope' )
	commands.append( gencommand + ' pdfShapeVar pdfShapeEnv envelope' )
	commands.append( gencommand + ' pdfShapeVar pdfShapeRMS RMS' )
	for command in commands: os.system(command)

    # step 4: clip histograms (if not fake rate)
    for f in filestomerge:
	if 'fakerate' in f.split('/')[-2]: continue
	print('now clipping {}'.format(f))
	histtools.clipallhistograms(f,mustcontain=['_rebinnedeventBDT'])	

if __name__=='__main__':

    sys.stderr.write('###starting###\n')

    # command line arguments: a top-level directory and the rebinning parameters
    # (not included in command line args: files to run on, see below)
    # a specific year, region and npmode can also be specified
    # (if not, it will run on all years and regions defined below)
    # note: append 'f' to firstbincount to indicate fraction of total events instead of absolute!
    topdir = ''
    mode = ''
    firstbincount = ''
    binfactor = ''
    year = ''
    region = ''
    npmode = ''
    runmultiple = True
    if len(sys.argv)==5:
	topdir = sys.argv[1]
	mode = sys.argv[2]
	firstbincount = sys.argv[3]
	binfactor = sys.argv[4]
    elif len(sys.argv)==8:
	topdir = sys.argv[1]
	mode = sys.argv[2]
        firstbincount = sys.argv[3]
        binfactor = sys.argv[4]
	year = sys.argv[5]
	region = sys.argv[6]
	npmode = sys.argv[7]
	runmultiple = False
    else:
        print('### ERROR ###: rebinoutput.py needs three command line arguments.')
        print('           normal usage: rebinoutput <output_directory> <mode> ')
	print('                         <first_bin_count> <bin_factor>')
        sys.exit()

    # global setting of years to run on
    yearlist = []
    yearlist.append('2016')
    yearlist.append('2017')
    yearlist.append('2018')
    
    # global setting of regions to run on
    regionlist = []
    regionlist.append('signalregion_1')
    for r in ['signalregion_2','signalregion_3']: regionlist.append(r)
    # recommended to not rebin the control regions, it is useless anyway 
    # and can give problems if using signal-based rebinning.
    #regionlist.append('wzcontrolregion')
    #regionlist.append('zzcontrolregion')
    #regionlist.append('zgcontrolregion')
    #for r in ['signalsideband_noossf_1']: regionlist.append(r)
    #for r in ['signalsideband_noz_1']: regionlist.append(r)

    # global setting of nonprompt mode to run on
    # WARNING: do not run on multiple modes simultaneously as the same files 
    # might be affected simultaneuously in a different way!
    # maybe try to fix later, e.g. by copying all files per element in the loop
    # to a different directory (but this comes with other disadvantages for the next steps)
    npmode = 'npfromdata'

    # check preconditions
    if not os.path.exists('./rebinoutput'):
	print('### ERROR ###: rebinoutput executable does not seem to exist')
	sys.exit()
    if not os.path.exists('./reenvelope'):
	print('### ERROR ###: reenvelope executable does not seem to exist')
        sys.exit()
    if mode not in ['signal','background']:
	print('### ERROR ###: mode "'+str(mode)+'" not recognized.')
        sys.exit()

    if runmultiple:
	cwd = os.getcwd()
	for year in yearlist:
	    for region in regionlist:
		script_name = 'rebinoutput.sh'
		with open(script_name,'w') as script:
		    initializeJobScript(script)
		    script.write('cd {}\n'.format(cwd))
		    command = 'python rebinoutput.py {} {} {} {} {} {} {}'.format(
				    topdir, mode, firstbincount, binfactor, year, region, npmode )
		    script.write(command+'\n')
		submitQsubJob(script_name)
		# alternative: run locally
		#os.system('bash '+script_name)

    else:
	rebinoutput(topdir,year,region,npmode,mode,firstbincount,binfactor)

    sys.stderr.write('###done###\n')
